src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionEmitMemberAccessor.cs

```diff
@@ -258,6 +258,10 @@ private static DynamicMethod CreatePropertyGetter(PropertyInfo propertyInfo, Typ
 
             if (declaredPropertyType != runtimePropertyType && declaredPropertyType.IsValueType)
             {
+                // Not supported scenario: possible if declaredPropertyType == int? and runtimePropertyType == int
```

Comment:
Alternatively, we could decide to support this scenario by doing something like:
```csharp
                if (runtimePropertyType.IsValueType)
                {
                    Debug.Assert(declaredPropertyType.IsNullableOfT() || runtimePropertyType.IsNullableOfT());
                    generator.Emit(OpCodes.Box, declaredPropertyType);
                    generator.Emit(OpCodes.Unbox_Any, runtimePropertyType);
                }
                else
                {
                    generator.Emit(OpCodes.Box, declaredPropertyType);
                }
```

Reply:
I think it's better to disallow it as done in this PR and have users provide converter implementations that align with what we intend. This way we can keep IL logic as simplified as possible and reduce the possibility of type errors.

- use a `JsonConverter<double>` if you want custom support for `double` and `double?` (supported by the serializer wrapping custom converter in `NullableConverter<>`)
- use a `JsonConverter<double?>` if you want custom support for `double?` specifically (for whatever reason). A separate `JsonConverter<double>` will be needed for `double` support in this case.

Reply:
With your fix, do we still have coverage for this branch? Same question for the location of the new debug assert below.

Reply:
Yeah, they're both covered by the new tests I added.